packages/core/src/lib/history.ts

```diff
@@ -0,0 +1,110 @@
+import { LogResult, simpleGit } from 'simple-git';
+import {
+  CoreConfig,
+  Format,
+  PersistConfig,
+  UploadConfig,
+  uploadConfigSchema,
+} from '@code-pushup/models';
+import { getCurrentBranchOrTag, safeCheckout } from '@code-pushup/utils';
+import { collectAndPersistReports } from './collect-and-persist';
+import { GlobalOptions } from './types';
+import { upload as uploadCommandLogic } from './upload';
+
+export type HistoryOnlyOptions = {
+  targetBranch?: string;
+  uploadReports?: boolean;
+  forceCleanStatus?: true;
+};
+export type HistoryOptions = Required<
+  Pick<CoreConfig, 'plugins' | 'categories'> & {
+    persist: Required<PersistConfig>;
+  } & { upload: Required<UploadConfig> }
+> &
+  GlobalOptions &
+  HistoryOnlyOptions;
+
+export async function history(
+  config: HistoryOptions,
+  commits: string[],
+): Promise<string[]> {
+  const initialBranch: string = await getCurrentBranchOrTag();
+
+  const { uploadReports = true } = config as unknown as HistoryOptions;
+  if (!uploadReports) {
+    console.warn('Upload is skipped because uploadReports is set to false');
+  }
+
+  const reports: string[] = [];
+  // eslint-disable-next-line functional/no-loop-statements
+  for (const commit of commits) {
+    console.info(`Collect ${commit}`);
+    await safeCheckout(commit, { forceCleanStatus: config.forceCleanStatus });
+
+    const currentConfig: HistoryOptions = {
+      ...config,
+      persist: {
+        ...config.persist,
+        format: ['json' as Format],
+        filename: `${commit}-report`,
+      },
+    };
+
+    await collectAndPersistReports(currentConfig);
+
+    if (uploadReports) {
+      const result = uploadConfigSchema.safeParse(currentConfig.upload);
+      if (result.success) {
+        await uploadCommandLogic({ ...currentConfig, upload: result.data });
+      } else {
+        console.error(`Collecting ${commit} failed.`);
+        console.error(result.error);
+      }
+    }
+
+    // eslint-disable-next-line functional/immutable-data
+    reports.push(currentConfig.persist.filename);
+  }
+
+  await safeCheckout(initialBranch, {
+    forceCleanStatus: config.forceCleanStatus,
+  });
+
+  return reports;
+}
+
+export async function getHashes(
+  options: {
+    from?: string;
+    to?: string;
+    maxCount?: number;
+  } = {},
+  git = simpleGit(),
+): Promise<string[]> {
+  const { from, to, maxCount } = options;
+
+  if (from || to) {
+    // validate from & to
+    if (from === undefined || from === '') {
+      throw new Error('from has to be defined');
+    }
+    if (to === undefined || to === '') {
+      throw new Error('to has to be defined');
+    }
```

Comment:
Why require both? The `git log` command with only `--since` or only `--until`.

Reply:
I thought from-to is a range 🤷‍♂️

Reply:
It is a range, it's just that you don't need specify both ends.

![image](https://github.com/code-pushup/cli/assets/34691111/f3ffbef7-2939-4a96-b13e-beba83592de7)

Reply:
I `Pick<LogOption, 'maxCount' | 'from' | 'to'>` and will move validation into cli.

Reply:
Thx for the examples! did not know about [--since](https://git-scm.com/docs/git-log#Documentation/git-log.txt---sinceltdategt)

Reply:
Ah, I misunderstood what `from` and `to` mean in `simple-git`'s `log` method. They're not equivalent to `since` and `until`, but they're the positional args in `git log` (`git log <from>`, `git log <from>..<to>`) :man_facepalming:

In that case, the updated validation makes sense :+1: